### DISCUSSION TITLE
Does openfl collaborators train sequentially?

### DISCUSSION THREAD

**Discussion by chirag2796** [securefederatedai/openfederatedlearning#421] [General]
Hi, I wanted to know if the openfl collaborators train one by one in a round before aggregation, or does it happen at the same time in parallel like a real FL setting, and the aggregator waits for the last collaborator to complete the round training?

I think it is the former, but just asking for my sanity sake. If it does happen one by one, is there any way to make it (or simulate it) parallel. I am doing some baseline experiments for a research and need to count parallel time. Any unconventional/hack way of doing it is also appreciated.

Thanks.

**Comment by psfoley** [comment]
Welcome to the community, @chirag2796, and thanks for using OpenFL!

Yes, using the [Aggregator-based workflow](https://openfl.readthedocs.io/en/latest/running_the_federation.html#aggregator-based-workflow) or [Director-based workflow](https://openfl.readthedocs.io/en/latest/running_the_federation.html#director-based-workflow) the collaborators will all train in parallel. The only interface where collaborators train sequentially is through our python API ([example here](https://github.com/intel/openfl/blob/develop/openfl-tutorials/Federated_Pytorch_MNIST_Tutorial.ipynb)). If you are currently using that interface, it should be pretty straightforward to migrate your code (model, dataloader, etc.) to the director-based workflow.
